google/cloud/storage/internal/hash_function_impl_test.cc

```diff
@@ -35,82 +38,191 @@ auto constexpr kEmptyStringMD5Hash = "1B2M2Y8AsgTpgAmY7PhCfg==";
+    auto const message_crc = storage_internal::Crc32c(message);
+    EXPECT_STATUS_OK(function.Update(npos, message, message_crc));
+    EXPECT_THAT(function.Update(npos, message, message_crc),
+                StatusIs(StatusCode::kInvalidArgument));
+  }
+  auto const actual = function.Finish();
+  EXPECT_EQ(actual.crc32c, kQuickFoxCrc32cChecksum);
+  EXPECT_THAT(actual.md5, IsEmpty());
+
+  auto const a2 = function.Finish();
+  EXPECT_EQ(a2.crc32c, kQuickFoxCrc32cChecksum);
+  EXPECT_THAT(a2.md5, IsEmpty());
+}
+
+TEST(HashFunctionImplTest, Crc32cCord) {
+  Crc32cHashFunction function;
+  auto payload = absl::Cord(absl::string_view{kQuickFox});
+  for (std::size_t npos = 0; npos < payload.size(); npos += 5) {
+    auto message = payload.Subcord(npos, 5);
+    auto const message_crc = storage_internal::Crc32c(message);
+    EXPECT_STATUS_OK(function.Update(npos, message, message_crc));
+    EXPECT_THAT(function.Update(npos, message, message_crc),
+                StatusIs(StatusCode::kInvalidArgument));
+  }
+  auto const actual = function.Finish();
+  EXPECT_EQ(actual.crc32c, kQuickFoxCrc32cChecksum);
+  EXPECT_THAT(actual.md5, IsEmpty());
+
+  auto const a2 = function.Finish();
+  EXPECT_EQ(a2.crc32c, kQuickFoxCrc32cChecksum);
+  EXPECT_THAT(a2.md5, IsEmpty());
+}
+
+TEST(HashFunctionImplTest, MD5Empty) {
   MD5HashFunction function;
   auto result = std::move(function).Finish();
   EXPECT_THAT(result.crc32c, IsEmpty());
   EXPECT_THAT(result.md5, kEmptyStringMD5Hash);
 }
 
-TEST(HasFunctionImplTest, QuickMD5) {
+TEST(HashFunctionImplTest, MD5Quick) {
   MD5HashFunction function;
-  Update(function, "The quick");
-  Update(function, " brown");
-  Update(function, " fox jumps over the lazy dog");
+  function.Update("The quick");
+  function.Update(" brown");
+  function.Update(" fox jumps over the lazy dog");
   auto result = std::move(function).Finish();
   EXPECT_THAT(result.crc32c, IsEmpty());
   EXPECT_THAT(result.md5, kQuickFoxMD5Hash);
 }
 
-TEST(HasFunctionImplTest, EmptyComposite) {
+TEST(HashFunctionImplTest, MD5StringView) {
+  MD5HashFunction function;
+  auto payload = absl::string_view{kQuickFox};
+  for (std::size_t npos = 0; npos < payload.size(); npos += 5) {
+    auto message = payload.substr(npos, 5);
+    EXPECT_STATUS_OK(function.Update(npos, message));
+    EXPECT_THAT(function.Update(npos, message),
+                StatusIs(StatusCode::kInvalidArgument));
+  }
+  auto const actual = function.Finish();
+  EXPECT_THAT(actual.crc32c, IsEmpty());
+  EXPECT_THAT(actual.md5, kQuickFoxMD5Hash);
+
+  auto const a2 = function.Finish();
+  EXPECT_THAT(a2.crc32c, IsEmpty());
+  EXPECT_THAT(a2.md5, kQuickFoxMD5Hash);
+}
+
+TEST(HashFunctionImplTest, MD5StringViewWithCrc) {
+  MD5HashFunction function;
+  auto payload = absl::string_view{kQuickFox};
+  for (std::size_t npos = 0; npos < payload.size(); npos += 5) {
+    auto message = payload.substr(npos, 5);
+    auto const message_crc = storage_internal::Crc32c(message);
```

Comment:
this just gets ignored, right? I assume the MD5 and CRC hashes do not mix.

Might it be clearer to say something like: `std::uint32_t const ignored_crc = 0`; ?

(If the real implementation provides a crc32c to the MD5 `Update` function, then I am okay with the test as is.)

Reply:
renamed to `unused`.